Sources/Algorithms/Partition.swift

```diff
@@ -204,3 +204,171 @@ extension Collection {
+  ///     let (longNames, shortNames) = cast.partitioned({ $0.count < 5 })
+  ///     print(longNames)
+  ///     // Prints "["Vivien", "Marlon"]"
+  ///     print(shortNames)
+  ///     // Prints "["Kim", "Karl"]"
+  ///
+  /// - Parameter belongsInSecondCollection: A closure that takes an element of
+  /// the sequence as its argument and returns a Boolean value indicating
+  /// whether the element should be included in the second returned array.
+  /// Otherwise, the element will appear in the first returned array.
+  ///
+  /// - Returns: Two arrays with with all of the elements of the receiver. The
+  /// first array contains all the elements that `belongsInSecondCollection`
+  /// didn’t allow, and the second array contains all the elements that
+  /// `belongsInSecondCollection` allowed.
+  ///
+  /// - Complexity: O(*n*), where *n* is the length of the sequence.
+  ///
+  /// - Note: This algorithm performs a bit slower than the same algorithm on
+  /// `RandomAccessCollection` since the size of the sequence is unknown, unlike
+  /// `RandomAccessCollection`.
+  @inlinable
+  public func partitioned(
+    _ belongsInSecondCollection: (Element) throws -> Bool
+  ) rethrows -> ([Element], [Element]) {
+    return try _partitioned(belongsInSecondCollection)
+  }
+  
+  @inlinable
+  public func _partitioned(
+    _ belongsInSecondCollection: (Element) throws -> Bool
+  ) rethrows -> ([Element], [Element]) {
+    var lhs = ContiguousArray<Element>()
+    var rhs = ContiguousArray<Element>()
+    
+    for element in self {
+      if try belongsInSecondCollection(element) {
+        rhs.append(element)
+      } else {
+        lhs.append(element)
+      }
+    }
+    
+    return _tupleMap((lhs, rhs), { Array($0) })
+  }
+}
+
+extension Collection {
+  // This is a specialized version of the same algorithm on `Sequence` that
+  // avoids reallocation of arrays since `count` is known ahead of time.
+  @inlinable
+  public func partitioned(
+    _ belongsInSecondCollection: (Element) throws -> Bool
+  ) rethrows -> ([Element], [Element]) {
+    guard !self.isEmpty else {
+      return ([], [])
+    }
+    
+    // Since `RandomAccessCollection`s have known sizes (access to `count` is
+    // constant time, O(1)), we can allocate one array of size `self.count`,
+    // then insert items at the beginning or end of that contiguous block. This
+    // way, we don’t have to do any dynamic array resizing. Since we insert the
+    // right elements on the right side in reverse order, we need to reverse
+    // them back to the original order at the end.
+    
+    let count = self.count
+    
+    // The overhead of this implementation isn’t worth it for collections fewer
+    // than 8 elements. Use the simple `Sequence`-based implementation instead.
+    // This constant was determined using benchmarking. More information:
+    // https://github.com/apple/swift-algorithms/pull/152#issuecomment-887130149
+    if count < 8 {
```

Comment:
Curious to hear feedback about how/where to define these kinds of constants.

Reply:
Using a literal alongside a comment on what it represents is fine. I'm not fully convinced we need this branch though, small cases are going to be very fast either way and it adds a bit of complexity. It also currently causes all tests to use the `Sequence` implementation 😅